### PR TITLE
Use Unsafe.SizeOf to calculate object sizes

### PR DESCRIPTION
**Pull request by PaulusParssinen**
Fixes potential GC hole by using BCL's `Unsafe.SizeOf<T>()` to calculate sizes for given type parameter more _safely_ 😆.

Contributes to #38

### CODE REVIEW

**Comment by PaulusParssinen** [comment]
No idea why build is failing in the CI..

File: libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs [hunk 1]
```diff
@@ -17,10 +17,10 @@ internal sealed unsafe class BlittableAllocator<Key, Value> : AllocatorBase<Key,
         private readonly long* nativePointers;
 
         // Record sizes
-        private static readonly int recordSize = Utility.GetSize(default(Record<Key, Value>));
-        private static readonly int recordInfoSize = Utility.GetSize(default(RecordInfo));
-        private static readonly int keySize = Utility.GetSize(default(Key));
-        private static readonly int valueSize = Utility.GetSize(default(Value));
+        private static readonly int recordSize = Unsafe.SizeOf<Record<Key, Value>>();
```
Comment (jkotas) [inline libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs:20]: It is unnecessary to cache these values in statics. `Unsafe.SizeOf` evaluates to constant, so it is cheaper to call it inline or to wrap it in a property.

File: libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs [hunk 2]
```diff
@@ -13,20 +13,17 @@ internal struct RevivificationManager<Key, Value>
         internal RevivificationStats stats = new();
 
         internal readonly bool IsEnabled = false;
-        internal readonly int FixedValueLength;
+        internal readonly int FixedValueLength => Unsafe.SizeOf<Value>();
         internal bool restoreDeletedRecordsIfBinIsFull;
         internal bool useFreeRecordPoolForCTT;
 
-        internal readonly bool IsFixedLength => FixedValueLength != 0;
+        internal readonly bool IsFixedLength { get; }
```
Comment (jkotas) [inline libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: Is this change needed? If original implementation would be more efficient.

**Comment by badrishc** [comment]
> ~No idea why build is failing in the CI..~
> 
> nvm. There apparently was a Tsavorite.test project that is _not_ visible in the normal project solution view. Fixing..

Yes, there are too many tests in Tsavorite.test and that degrades the dev experience for Garnet, so those tests are not included in the Garnet solution. They are limited to Tsavorite. Tsavorite CI kicks off if any changes to Tsavorite folder are detected.

File: libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs (hunk 2 above)
Comment (PaulusParssinen) [reply libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: The revivifaction logic was broken due to how isFixedLen was flowed in to this ctor. This was quickest fix I could come up to restore previous logic. Shouldn't JIT be able to consider this bool as constant in next tier as it is `readonly`?

Comment (PaulusParssinen) [reply libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: I wrecked my brain and couldn't come up how one of these variations would result less branching or worse CQ.. What am I missing?

Comment (jkotas) [reply libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: I agree. I think what you have is fine after taking a closer look.

File: libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs (hunk 1 above)
Comment (badrishc) [reply libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs]: is this true for .NET 6 as well?

File: libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs (hunk 2 above)
Comment (badrishc) [reply libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: @TedHartMS - pls check this thread as well.

File: libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs (hunk 1 above)
Comment (jkotas) [reply libs/storage/Tsavorite/cs/src/core/Allocator/BlittableAllocator.cs]: Yes, it is true for all .NET versions.

File: libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs (hunk 2 above)
Comment (TedHartMS) [reply libs/storage/Tsavorite/cs/src/core/Index/Tsavorite/Implementation/Revivification/RevivificationManager.cs:20]: This change fixes 9bdb889 that set FixedValueLength always, rather than setting it only if isFixedLen is true. While the fix in ef5a01a was sufficient to fix the GC hole, the change in 9bdb889 made it consistent with the pattern of calling Unsafe.SizeOf only in properties. This requires a separate bool value for IsFixedLength. Unfortunately, as you've discovered, this bool can't be considered a constant even though it is readonly.

**Review by TedHartMS — APPROVED** [review]
